src/ert/ensemble_evaluator/snapshot.py

```diff
@@ -104,110 +102,175 @@ def convert_iso8601_to_datetime(
+        else:
+            raise ValueError(
+                f"unknown step status {step_status} for real: {real_id} step: "
+                + f"{step_id}"
             )
-        )
         return self
 
     def to_dict(self) -> Mapping[str, Any]:
-        return cast(Mapping[str, Any], pyrsistent.thaw(self._data))
+        """used to send snapshot updates - for thread safety, this method should not
+        access the _snapshot property"""
+        _dict = {}
+        if self._metadata:
+            _dict["metadata"] = self._metadata
+        if self._ensemble_state:
+            _dict["status"] = self._ensemble_state
+        if self._realization_states:
+            _dict["reals"] = self._realization_states
+
+        for step_index_tuple, step_state in self._step_states.items():
+            real_id = step_index_tuple[0]
+            step_id = step_index_tuple[1]
+            if "reals" not in _dict:
+                _dict["reals"] = {real_id: {}}
+            if "steps" not in _dict["reals"][real_id]:
+                _dict["reals"][real_id]["steps"] = {}
+            if step_id not in _dict["reals"][real_id]["steps"]:
+                _dict["reals"][real_id]["steps"][step_id] = step_state
+                _dict["reals"][real_id]["steps"][step_id]["jobs"] = {}
+
+        for job_tuple, job_values_dict in self._job_states.items():
+            real_id = job_tuple[0]
+            step_id = job_tuple[1]
+            if "reals" not in _dict:
+                _dict["reals"] = {}
+            if real_id not in _dict["reals"]:
+                _dict["reals"][real_id] = {}
+            if "steps" not in _dict["reals"][real_id]:
+                _dict["reals"][real_id]["steps"] = {}
+            if step_id not in _dict["reals"][real_id]["steps"]:
+                _dict["reals"][real_id]["steps"][step_id] = {"jobs": {}}
+
+            job_id = job_tuple[2]
+            _dict["reals"][real_id]["steps"][step_id]["jobs"][job_id] = job_values_dict
+
+        return _dict
 
-    def data(self) -> TPMap[str, Any]:
-        return self._data
+    def data(self) -> Mapping[str, Any]:
+        return self.to_dict()
+
+    def _recursive_merge(self, other: "PartialSnapshot") -> "PartialSnapshot":
+        self._metadata.update(other._metadata)
+        if other._ensemble_state is not None:
+            self._ensemble_state = other._ensemble_state
+        for real_id, other_real_data in other._realization_states.items():
+            self._realization_states[real_id].update(other_real_data)
+        for step_id, other_step_data in other._step_states.items():
+            self._step_states[step_id].update(other_step_data)
+        for job_id, other_job_data in other._job_states.items():
+            self._job_states[job_id].update(other_job_data)
+        return self
 
     # pylint: disable=too-many-branches
     def from_cloudevent(self, event: CloudEvent) -> "PartialSnapshot":
+        # pylint: disable=too-many-statements   # TODO: fix this before merge
```

Comment:
I am okay with not fixing this before merge.

Reply:
removed comment